### PR TITLE
impl(common): TimerQueue supports per-timer cancellation

### PR DESCRIPTION
**Pull request by coryan**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10544)

### CODE REVIEW

File: google/cloud/internal/timer_queue_test.cc [hunk 1]
```diff
@@ -29,51 +29,68 @@ using ::google::cloud::testing_util::StatusIs;
 using testing::Each;
 
 TEST(TimerQueueTest, ScheduleSingleRunner) {
-  TimerQueue tq;
-  std::thread t([&tq] { tq.Service(); });
+  auto tq = TimerQueue::Create();
+  std::thread t([&tq] { tq->Service(); });
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/timer_queue_test.cc]: Done

File: google/cloud/internal/timer_queue.h [hunk 2]
```diff
@@ -80,18 +84,21 @@ class TimerQueue {
   // call Service. Calls to Service only return after Shutdown has been called.
   void Service();
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/timer_queue.h:85]: Removed

File: google/cloud/internal/timer_queue.cc [hunk 3]
```diff
@@ -104,27 +112,31 @@ void TimerQueue::Service() {
...
-void TimerQueue::CancelAll() {
-  CancelAll(std::unique_lock<std::mutex>(mu_), "TimerQueue cancel all");
-}
-
-void TimerQueue::CancelAll(std::unique_lock<std::mutex> lk, char const* msg) {
   while (!timers_.empty()) {
     auto p = std::move(timers_.begin()->second);
     timers_.erase(timers_.begin());
     lk.unlock();
     cv_.notify_one();
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/timer_queue.cc]: Removed

File: google/cloud/internal/timer_queue.h [hunk 4]
```diff
@@ -80,18 +84,21 @@ class TimerQueue {
...
-  // Requires a lock to be held before calling.
-  void CancelAll(std::unique_lock<std::mutex> lk, char const* msg);
-
   using PromiseType = promise<StatusOr<std::chrono::system_clock::time_point>>;
+  using KeyType =
+      std::pair<std::chrono::system_clock::time_point, std::uint64_t>;
+
+  TimerQueue() = default;
+
+  // Cancels an specific timer
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/timer_queue.h]: Fixed.

File: google/cloud/internal/timer_queue.cc [hunk 5]
```diff
@@ -104,27 +113,30 @@ void TimerQueue::Service() {
     // This may run user code (in the continuations for the future). That may
     // do all kinds of things, including calling back into this class to create
     // new timers. We cannot hold the mutex while it is running.
-    p.set_value(until);
+    p.set_value(key.first);
```
Comment (devbww) [inline google/cloud/internal/timer_queue.cc:116]: The test expectations can now be made exact.

File: google/cloud/internal/timer_queue.h [hunk 6]
```diff
@@ -62,36 +63,42 @@ namespace internal {
...
+  TimerQueue(TimerQueue&&) = delete;
+  TimerQueue& operator=(TimerQueue&&) = delete;
 
   // Adds a timer to the queue.
   future<StatusOr<std::chrono::system_clock::time_point>> Schedule(
       std::chrono::system_clock::time_point tp);
 
   // Signals all threads that have called Service to return. Does not modify
-  // remaining timers. If desired, CancelAll can be called.
+  // remaining timers.
```
Comment (devbww) [inline google/cloud/internal/timer_queue.h:80]: "Does not modify remaining timers" is misleading, given than this results in them being cancelled.

**Review by devbww — APPROVED** [review]

File: google/cloud/internal/timer_queue.h (hunk 6 above)
Comment (coryan) [reply google/cloud/internal/timer_queue.h]: A made a larger pass through these comments.  Could you take a look?

File: google/cloud/internal/timer_queue.cc (hunk 5 above)
Comment (coryan) [reply google/cloud/internal/timer_queue.cc:116]: Done.

File: google/cloud/internal/timer_queue_test.cc [hunk 7]
```diff
@@ -80,7 +80,7 @@ TEST(TimerQueueTest, ScheduleEarlierTimerSingleRunner) {
   auto tq = TimerQueue::Create();
   std::thread t([tq] { tq->Service(); });
   auto const duration = std::chrono::milliseconds(50);
-  auto now = std::chrono::system_clock::now();
+  auto const now = std::chrono::system_clock::now();
```
Comment (devbww) [inline google/cloud/internal/timer_queue_test.cc:83]: "const now" is blowing my mind.  :-)

File: google/cloud/internal/timer_queue.h (hunk 6 above)
Comment (devbww) [reply google/cloud/internal/timer_queue.h]: Nice.

**Review by devbww — APPROVED** [review]

**Pull request merged**